lib/AST/Type.cpp

```diff
@@ -157,22 +157,73 @@ bool TypeBase::isMarkerExistential() {
   return true;
 }
 
-bool TypeBase::isNoncopyable() {
-  if (auto *nom = getAnyNominal())
-    return nom->isMoveOnly();
+/// Returns true if this type is _always_ Copyable using the legacy check
+/// that does not rely on conformances.
+static bool alwaysNoncopyable(Type ty) {
+  if (auto *nominal = ty->getNominalOrBoundGenericNominal())
+    return nominal->isNoncopyable();
 
-  if (auto *expansion = getAs<PackExpansionType>()) {
-    return expansion->getPatternType()->isNoncopyable();
+  if (auto *expansion = ty->getAs<PackExpansionType>()) {
+    return alwaysNoncopyable(expansion->getPatternType());
   }
 
   // if any components of the tuple are move-only, then the tuple is move-only.
-  if (auto *tupl = getCanonicalType()->getAs<TupleType>()) {
+  if (auto *tupl = ty->getCanonicalType()->getAs<TupleType>()) {
     for (auto eltTy : tupl->getElementTypes())
-      if (eltTy->isNoncopyable())
+      if (alwaysNoncopyable(eltTy))
         return true;
   }
 
-  return false;
+  return false; // otherwise, the conservative assumption is it's copyable.
+}
+
+/// \returns true iff this type lacks conformance to Copyable, using the given
+/// context to substitute unbound types.
+bool TypeBase::isNoncopyable(const DeclContext *dc) {
+  assert(dc);
+
+  /** FIXME: seems this is busted :(
+  // Fast-path for type parameters; ask the generic signature directly and
+  // cache that answer.
+  if (isTypeParameter()) {
+    auto canType = getCanonicalType();
+    auto &ctx = canType->getASTContext();
+    IsNoncopyableRequest req {canType};
```

Comment:
The generic signature is not part of the caching key so that might be why it's busted. However it's not worth caching GenericSignature::requiresProtocol(); it's extremely fast.

Reply:
Yeah removing the caching fixed this! I wasn't sure if the call into the requirement machine was fast enough to be used uncached.